core/src/org/labkey/core/notification/NotificationEndpoint.java

```diff
@@ -189,20 +206,38 @@ private static void doClose(int userId, @Nullable HttpSession httpSession, Strin
             LOG.debug("WebSocket: no sessions to close for " + userId + " (" + (httpSession != null ? httpSession.getId() : "all sessions") + "): " + message);
     }
 
-
     private static void sendEvent(int userId, String eventName)
+    {
+        List<Integer> users = new ArrayList<>();
+        users.add(userId);
+
+        sendEvent(users, eventName);
+    }
+
+    private static void sendEvent(List<Integer> userIds, String eventName)
     {
         final String data = "{\"event\":\"" + eventName + "\"}";
-        long count = endpoints(userId)
-                .stream()
-                .map(e -> e.safely(() -> {
-                    e.session.getBasicRemote().sendText(data);
-                    LOG.debug(e.toString() + " sendText: " + eventName);
-                }))
-                .count();
+        List<NotificationEndpoint> endpoints = endpoints(userIds);
+
+        long count = 0;
+        // do not refactor to use stream as it might result in deadlock
+        for (NotificationEndpoint endpoint : endpoints)

```

Comment:
Looks like you could just inline the call to endpoints here:
```suggestion
        for (NotificationEndpoint endpoint : endpoints(userIds))

```